File: storage/db-core/src/main/java/com/coffee/gu/menu/MenuEntity.java
package com.coffee.gu.menu;

import com.coffee.gu.enums.MenuType;
import com.coffee.gu.BaseEntity;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;

import java.math.BigDecimal;

@Table(name = "menu")
@Entity
public class MenuEntity extends BaseEntity {

    private String name;
    @Enumerated(EnumType.STRING)
    private MenuType type;
    private BigDecimal costPrice;
    private BigDecimal salesPrice;
    private String imageUrl;
    private String description;
    private Double capacity;
    private Double caffeine;
    private Double calories;
    private Double sodium;
    private Double carbohydrate;
    private Double sugar;
    private Double fat;
    private Double saturatedFat;
    private Double protein;
    private String containedAllergens;
    private String mayContainAllergens;

    public MenuEntity() {}

    private MenuEntity(String name, MenuType type, BigDecimal costPrice, BigDecimal salesPrice, String imageUrl, String description, Double capacity, Double caffeine, Double calories, Double sodium, Double carbohydrate, Double sugar, Double fat, Double saturatedFat, Double protein, String containedAllergens, String mayContainAllergens) {
        this.name = name;
        this.type = type;
        this.costPrice = costPrice;
        this.salesPrice = salesPrice;
        this.imageUrl = imageUrl;
        this.description = description;
        this.capacity = capacity;
        this.caffeine = caffeine;
        this.calories = calories;
        this.sodium = sodium;
        this.carbohydrate = carbohydrate;
        this.sugar = sugar;
        this.fat = fat;
        this.saturatedFat = saturatedFat;
        this.protein = protein;
        this.containedAllergens = containedAllergens;
        this.mayContainAllergens = mayContainAllergens;
    }

    public static MenuEntity from(Menu menu) {
        return new MenuEntity(
                menu.getName(),
                menu.getType(),
                menu.getCostPrice(),
                menu.getSalesPrice(),
                menu.getImageUrl(),
                menu.getDescription(),
                menu.getDetail().nutrition().capacity(),
                menu.getDetail().nutrition().caffeine(),
                menu.getDetail().nutrition().calories(),
                menu.getDetail().nutrition().sodium(),
                menu.getDetail().nutrition().carbohydrate(),
                menu.getDetail().nutrition().sugar(),
                menu.getDetail().nutrition().fat(),
                menu.getDetail().nutrition().saturatedFat(),
                menu.getDetail().nutrition().protein(),
                menu.getDetail().containedAllergens(),
                menu.getDetail().mayContainAllergens()
        );
    }

    public Menu toModel() {
        return new Menu(
                this.id,
                this.name,
                this.type,
                new Price(this.costPrice, this.salesPrice),
                this.imageUrl,
                this.description,
                new MenuDetail(
                        new Nutrition(
                                this.capacity,
                                this.caffeine,
                                this.calories,
                                this.sodium,
                                this.carbohydrate,
                                this.sugar,
                                this.fat,
                                this.saturatedFat,
                                this.protein
                        ),
                        this.containedAllergens,
                        this.mayContainAllergens
                )
        );
    }

    public String getName() {
        return name;
    }
    public MenuType getType() {
        return type;
    }
    public BigDecimal getCostPrice() {
        return costPrice;
    }
    public BigDecimal getSalesPrice() {
        return salesPrice;
    }
    public String getImageUrl() { return imageUrl; }
    public String getDescription() { return description; }

    public Double getCapacity() {
        return capacity;
    }

    public Double getCaffeine() {
        return caffeine;
    }

    public Double getCalories() {
        return calories;
    }

    public Double getSodium() {
        return sodium;
    }

    public Double getCarbohydrate() {
        return carbohydrate;
    }

    public Double getSugar() {
        return sugar;
    }

    public Double getFat() {
        return fat;
    }

    public Double getSaturatedFat() {
        return saturatedFat;
    }

    public Double getProtein() {
        return protein;
    }

    public String getContainedAllergens() {
        return containedAllergens;
    }

    public String getMayContainAllergens() {
        return mayContainAllergens;
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/menu/MenuEntity.kt
package com.coffee.gu.menu

import com.coffee.gu.BaseEntity
import com.coffee.gu.enums.MenuType
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.math.BigDecimal

@Table(name = "menu")
@Entity
class MenuEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    val name: String,
    @Enumerated(EnumType.STRING)
    val type: MenuType,
    val costPrice: BigDecimal,
    val salesPrice: BigDecimal,
    val imageUrl: String? = null,
    val description: String? = null,
    val capacity: Double? = null,
    val caffeine: Double? = null,
    val calories: Double? = null,
    val sodium: Double? = null,
    val carbohydrate: Double? = null,
    val sugar: Double? = null,
    val fat: Double? = null,
    val saturatedFat: Double? = null,
    val protein: Double? = null,
    val containedAllergens: String? = null,
    val mayContainAllergens: String? = null,
) : BaseEntity() {

    fun toModel(): Menu {
        return Menu(
            id = id,
            name = name,
            type = type,
            price = Price(costPrice, salesPrice),
            imageUrl = imageUrl,
            description = description,
            detail = MenuDetail(
                nutrition = Nutrition(
                    capacity = capacity,
                    caffeine = caffeine,
                    calories = calories,
                    sodium = sodium,
                    carbohydrate = carbohydrate,
                    sugar = sugar,
                    fat = fat,
                    saturatedFat = saturatedFat,
                    protein = protein,
                ),
                containedAllergens = containedAllergens,
                mayContainAllergens = mayContainAllergens,
            ),
        )
    }

    companion object {
        fun from(menu: Menu): MenuEntity = MenuEntity(
            id = menu.id,
            name = menu.name ?: "",
            type = menu.type ?: MenuType.DRINK,
            costPrice = menu.price?.costPrice ?: BigDecimal.ZERO,
            salesPrice = menu.price?.salesPrice ?: BigDecimal.ZERO,
            imageUrl = menu.imageUrl,
            description = menu.description,
            capacity = menu.detail?.nutrition?.capacity,
            caffeine = menu.detail?.nutrition?.caffeine,
            calories = menu.detail?.nutrition?.calories,
            sodium = menu.detail?.nutrition?.sodium,
            carbohydrate = menu.detail?.nutrition?.carbohydrate,
            sugar = menu.detail?.nutrition?.sugar,
            fat = menu.detail?.nutrition?.fat,
            saturatedFat = menu.detail?.nutrition?.saturatedFat,
            protein = menu.detail?.nutrition?.protein,
            containedAllergens = menu.detail?.containedAllergens,
            mayContainAllergens = menu.detail?.mayContainAllergens,
        )
    }
}
